src/modules/schnorrsig/main_impl.h

```diff
@@ -0,0 +1,360 @@
+        noncefp = secp256k1_nonce_function_bip340;
+    }
+    secp256k1_scalar_set_b32(&x, seckey, &overflow);
+    /* Fail if the secret key is invalid. */
+    if (overflow || secp256k1_scalar_is_zero(&x)) {
+        memset(sig, 0, sizeof(*sig));
+        return 0;
+    }
+
+    secp256k1_ecmult_gen(&ctx->ecmult_gen_ctx, &pkj, &x);
+    secp256k1_ge_set_gej(&pk, &pkj);
+
+    /* Because we are signing for a x-only pubkey, the secret key is negated
+     * before signing if the point corresponding to the secret key does not
+     * have a square Y. */
+    secp256k1_fe_normalize(&pk.y);
+    if (secp256k1_fe_is_odd(&pk.y)) {
+        secp256k1_scalar_negate(&x, &x);
+    }
+
+    secp256k1_scalar_get_b32(seckey_tmp, &x);
+    secp256k1_fe_normalize(&pk.x);
+    secp256k1_fe_get_b32(pk_buf, &pk.x);
+    if (!noncefp(buf, msg32, seckey_tmp, pk_buf, (unsigned char *) "BIP340/nonce0000", (void*)ndata, 0)) {
+        memset(sig, 0, sizeof(*sig));
+        memset(seckey_tmp, 0, sizeof(seckey_tmp));
+        secp256k1_scalar_clear(&x);
+        return 0;
+    }
+    memset(seckey_tmp, 0, sizeof(seckey_tmp));
+
+    secp256k1_scalar_set_b32(&k, buf, NULL);
+    if (secp256k1_scalar_is_zero(&k)) {
+        memset(sig, 0, sizeof(*sig));
+        secp256k1_scalar_clear(&x);
+        return 0;
+    }
+
+    secp256k1_ecmult_gen(&ctx->ecmult_gen_ctx, &rj, &k);
+    secp256k1_ge_set_gej(&r, &rj);
+
+    if (!secp256k1_fe_is_quad_var(&r.y)) {
+        secp256k1_scalar_negate(&k, &k);
+    }
+    secp256k1_fe_normalize(&r.x);
+    secp256k1_fe_get_b32(&sig->data[0], &r.x);
+
+    /* tagged hash(r.x, pk.x, msg32) */
+    secp256k1_schnorrsig_sha256_tagged(&sha);
+    secp256k1_sha256_write(&sha, &sig->data[0], 32);
+    secp256k1_sha256_write(&sha, pk_buf, sizeof(pk_buf));
+    secp256k1_sha256_write(&sha, msg32, 32);
+    secp256k1_sha256_finalize(&sha, buf);
+
+    secp256k1_scalar_set_b32(&e, buf, NULL);
+    secp256k1_scalar_mul(&e, &e, &x);
+    secp256k1_scalar_add(&e, &e, &k);
+
+    secp256k1_scalar_get_b32(&sig->data[32], &e);
+    secp256k1_scalar_clear(&k);
+    secp256k1_scalar_clear(&x);
+
+    return 1;
+}
+
+int secp256k1_schnorrsig_verify(const secp256k1_context* ctx, const secp256k1_schnorrsig *sig, const unsigned char *msg32, const secp256k1_xonly_pubkey *pubkey) {
+    secp256k1_scalar s;
+    secp256k1_scalar e;
+    secp256k1_gej rj;
+    secp256k1_ge pk;
+    secp256k1_gej pkj;
+    secp256k1_fe rx;
+    secp256k1_sha256 sha;
+    unsigned char buf[32];
+    int overflow;
+
+    VERIFY_CHECK(ctx != NULL);
+    ARG_CHECK(secp256k1_ecmult_context_is_built(&ctx->ecmult_ctx));
+    ARG_CHECK(sig != NULL);
+    ARG_CHECK(msg32 != NULL);
+    ARG_CHECK(pubkey != NULL);
+
+    if (!secp256k1_fe_set_b32(&rx, &sig->data[0])) {
+        return 0;
+    }
+
+    secp256k1_scalar_set_b32(&s, &sig->data[32], &overflow);
+    if (overflow) {
+        return 0;
+    }
+
+    if (!secp256k1_xonly_pubkey_load(ctx, &pk, pubkey)) {
```

Comment:
Is `secp256k1_xonly_pubkey_load`  / `secp256k1_pubkey_load` an equivalent of `point(pk)` in BIP-340?
maybe it'll be good to add a reminder comment of what happens if `pubkey's x > p - 1` (if it fails here, as in BIP-340, what makes it fail /  `secp256k1_fe_from_storage` or `secp256k1_fe_set_b32`...?  -- or if it goes through and fails at the end, why it is fine)

Reply:
No, the equivalent is `secp256k1_xonly_pubkey_load` (the `pubkey` argument to `secp256k1_schnorrsig_verify` is already an `xonly_pubkey` while the BIP-340 `pk` is a byte array,).